### ISSUE TITLE
Slow decoding

### ISSUE THREAD

**Issue by eribean** [Rich-Hall/sentinel1decoder#3]
The decoding of the packet data is extremely slow for later pulses. From the data in your notebook, below is the time difference for reading two pulses at the beginning and toward the end of the file.

![image](https://user-images.githubusercontent.com/15330439/224582962-6f51fc86-7e9b-4a8e-a4a2-afce48ed7ca7.png)

I profiled the function ```decode_file``` and it is spending 97% in this line:

```python
if (input_header == this_header).all(1).any():
```

I looks like a you're comparing a dictionary to a data frame and require all values to match. I'm not very familiar with the inner workings of the raw data, but perhaps looking at the index is sufficient.

```python
if (any(input_header['Space Packet Count'] == this_header['Space Packet Count'])):
```

results in faster reads. I didn't spend a ton of time on this but maybe you know something more robust.

![image](https://user-images.githubusercontent.com/15330439/224583298-eab4d347-4676-4ff1-92d3-39a8fdfaae79.png)

Thanks~

**Comment by Rich-Hall** [comment]
Investigating this! The index *might* be sufficient, or alternatively perhaps we can hash some/all of the important values upon metadata decode. Stay tuned.

**Comment by Rich-Hall** [comment]
I replaced this with `if this_header['Space Packet Count'] in input_header['Space Packet Count'].values:` - let me know if this works better!

**Comment by eribean** [comment]
Thanks Rich, I'll provide an update as soon as I get back to the office.

**Comment by Rich-Hall** [comment]
If it helps, I've just pushed a commit that allows one to cache decoded bursts in an `.npy` file. On my machine although the initial decoding of a burst is still slow, subsequently the information can be recalled in seconds. The functionality is all wrapped in a new `Level0File` class.

On the demo notebook I've also now parallelized most functions which greatly speeds up that side of the image formation process.
